File: crypto.py
"""
AES-256-GCM authenticated encryption for VPN packets.

We use a counter-based nonce (8 bytes counter + 4 bytes random) to prevent
nonce reuse and detect replays. Key is derived from password via scrypt.
"""

import os
import struct
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.scrypt import Scrypt


class CryptoError(Exception):
    """Raised on decrypt/auth failure."""


class Cipher:
    """
    AES-256-GCM with embedded nonce.

    Wire format produced by encrypt():
        [ 12-byte nonce | ciphertext | 16-byte GCM auth tag ]

    The 12-byte nonce is sent in the clear (this is standard for AEAD ciphers;
    the auth tag protects against tampering).
    """

    NONCE_SIZE = 12
    KEY_SIZE = 32  # AES-256

    def __init__(self, key: bytes):
        if len(key) != self.KEY_SIZE:
            raise ValueError(f"Key must be {self.KEY_SIZE} bytes (AES-256)")
        self._aead = AESGCM(key)
        # Random session prefix + monotonic counter for unique nonces
        self._nonce_prefix = os.urandom(4)
        self._counter = 0

    def _next_nonce(self) -> bytes:
        self._counter += 1
        return self._nonce_prefix + struct.pack(">Q", self._counter)

    def encrypt(self, plaintext: bytes, aad: bytes = b"") -> bytes:
        nonce = self._next_nonce()
        ct = self._aead.encrypt(nonce, plaintext, aad)
        return nonce + ct

    def decrypt(self, data: bytes, aad: bytes = b"") -> bytes:
        if len(data) < self.NONCE_SIZE + 16:
            raise CryptoError("Packet too short to contain nonce + tag")
        nonce, ct = data[: self.NONCE_SIZE], data[self.NONCE_SIZE :]
        try:
            return self._aead.decrypt(nonce, ct, aad)
        except Exception as e:
            raise CryptoError(f"Decryption failed: {e}") from e


def derive_key(password: str, salt: bytes = b"vpn-project-v1-salt") -> bytes:
    """
    Derive a 32-byte key from a password using scrypt.

    Both client and server must use the *same* password and salt. In a real
    deployment, prefer a long random shared secret (32 bytes hex) over a
    human-typed password.
    """
    if not password:
        raise ValueError("Password cannot be empty")
    kdf = Scrypt(salt=salt, length=Cipher.KEY_SIZE, n=2**14, r=8, p=1)
    return kdf.derive(password.encode("utf-8"))

File: android/app/src/main/java/com/vpn/Crypto.kt
package com.vpn

import org.bouncycastle.crypto.generators.SCrypt
import java.nio.ByteBuffer
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * AES-256-GCM + scrypt KDF — مطابق تماماً لتطبيق Python.
 *
 * Wire format (encrypt output):
 *   [ 12-byte nonce | ciphertext | 16-byte GCM tag ]
 *
 * Nonce = 4-byte random prefix (per session) + 8-byte big-endian counter
 * Key   = scrypt(password, salt="vpn-project-v1-salt", n=2^14, r=8, p=1, len=32)
 */
object CryptoEngine {

    private val SALT = "vpn-project-v1-salt".toByteArray(Charsets.UTF_8)

    fun deriveKey(password: String): ByteArray =
        SCrypt.generate(password.toByteArray(Charsets.UTF_8), SALT, 16384, 8, 1, 32)

    class SessionCipher(key: ByteArray) {
        private val keySpec = SecretKeySpec(key, "AES")
        private val noncePrefix = ByteArray(4).also { SecureRandom().nextBytes(it) }
        private var counter = 0L

        private fun nextNonce(): ByteArray {
            counter++
            val counterBytes = ByteBuffer.allocate(8).putLong(counter).array() // big-endian
            return noncePrefix + counterBytes
        }

        fun encrypt(plaintext: ByteArray): ByteArray {
            val nonce = nextNonce()
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.ENCRYPT_MODE, keySpec, GCMParameterSpec(128, nonce))
            return nonce + cipher.doFinal(plaintext)
        }

        fun decrypt(data: ByteArray): ByteArray {
            require(data.size >= 28) { "Packet too short (${data.size} bytes)" }
            val nonce = data.copyOf(12)
            val ct = data.copyOfRange(12, data.size)
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.DECRYPT_MODE, keySpec, GCMParameterSpec(128, nonce))
            return cipher.doFinal(ct)
        }
    }
}
